### PR TITLE
Update UserInterfaceTests with template tests

### PR DESCRIPTION
**Pull request by manish**
Fixes UserInterfaceTests

Tests templates for templates in file `MonoDevelopTemplatesTest.cs`
* Console Project
* Gtk# 2.0 Project
* Library
* NUnit Library Project

Steps taken for all the template tests
* Open `New Solution` dialog
* Select `.NET`
* Select the template type
* Click `OK`
* Enter `Project Name`
* Enter `Solution Name`
* Enter `Location`
* Asserts that `Create a project within the solution directory` is enable and visible. If yes, then unchecks it.
* Asserts that `Create a .gitignore file to ignore inessential files` is enabled and visible. If yes, then unchecks it.
* calls `beforeBuild` action, which is the action to be taken before the solution is built. In case of NUnit Project, that delegate blocks waiting for `Package updates are available.` to be visible in Status Bar
* Builds the project
* Check if there any error
* Closes the solution
* Tries to delete the solution folder.

### CODE REVIEW

**Comment by manish** [comment]
@mhutch @iainx

**Comment by iainx** [comment]
approve

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Components.AutoTest/AutoTestSession.cs [hunk 1]
```diff
@@ -137,11 +138,32 @@ public void TypeText (string text)
...
+				var itemSelected = SelectWidget (treeViewName, true);
+				if (!itemSelected)
+					return false;
+
+				bool parentFound = after == null;
+				var treeView = currentObject as Gtk.TreeView;
+				bool result = false;
+				if (treeView != null) {
+					treeView.Model.Foreach ((model, path, iter) => {
+						var iterName = (string)model.GetValue (iter, 0);
```
Comment (slluis) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Components.AutoTest/AutoTestSession.cs]: We'll have to expose a semantic model of the tree. Item names are not always in column 0.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Components.AutoTest/AutoTestSession.cs [hunk 2]
```diff
@@ -406,7 +461,7 @@ public void WaitForWindow (string windowName, int timeout)
 
 			GLib.Timeout.Add ((uint) pollTime, () => {
 				var window = GetFocusedWindow (false);
-				if (window != null && window.Name == windowName) {
+				if (window != null && window.GetType ().FullName == windowName) {
```
Comment (slluis) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Components.AutoTest/AutoTestSession.cs]: The window type name is an implementation detail. We should avoid using those names in the tests.

File: main/tests/UserInterfaceTests/CreateBuildTemplatesTestBase.cs [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+			var solutionParentDirectory = Util.CreateTmpDir (projectName);
+			string actualSolutionDirectory = string.Empty;
+			try {
+				var newProject = new NewProjectController ();
+				newProject.Open ();
+
+				SelectTemplate (newProject, kind, category, categoryRoot);
+
+				Assert.IsTrue (newProject.Next ());
+				Thread.Sleep (3000);
```
Comment (slluis) [inline main/tests/UserInterfaceTests/CreateBuildTemplatesTestBase.cs]: Tests should not use fixed waits. Instead, they should wait for events from the UI.

File: main/tests/UserInterfaceTests/NewProjectController.cs [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+		public bool UseGit (bool useGit, bool useGitIgnore = true)
+		{
+			var gitSelectionSuccess = SetCheckBox ("useGitCheckBox", useGit);
+			if (gitSelectionSuccess && useGit)
+				return gitSelectionSuccess && SetCheckBox ("createGitIgnoreFileCheckBox", useGitIgnore);
+			return gitSelectionSuccess;
+		}
+
+		public bool SetCheckBox (string checkBoxName, bool active)
```
Comment (slluis) [inline main/tests/UserInterfaceTests/NewProjectController.cs]: This method and the others below are not specific to NewProjectController, so they should be in some other reusable class.

**Comment by slluis** [comment]
I added some comments about stuff that should be fixed in the test framework. However, I'm ok about committing this as is until we have a new testing infrastructure.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Components.AutoTest/AutoTestSession.cs (hunk 1 above)
Comment (manish) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Components.AutoTest/AutoTestSession.cs]: I have not yet proposed it in inclusion as I did not want it to drop huge amount of code, but I have it working in my private branch

**Pull request merged**